iplog.8
.\" $Id: iplog.8,v 1.1.1.1 2010/09/02 01:57:41 reynolds Exp $
.TH iplog 8 "03 December 2000" "iplog 2.2.3"

.SH NAME
iplog \- TCP/IP traffic logger.

.SH SYNOPSIS
.BI "iplog [options]"
.sp
.B "[\-DFILNPRSTUVbcdefhkmnopqstvwxyz]"
.br
.B "[\-a <network,network2,...>]"
.br
.B "[\-g <group>]"
.br
.B "[\-i <interface1,...,interfaceN>]"
.br
.B "[\-l <logfile>]"
.br
.B "[\-\-pid\-file=<file>]"
.br
.B "[\-u <user>]"
.br
.B "[\-\-tcp[=argument]]"
.br
.B "[\-\-udp[=argument]]"
.br
.B "[\-\-icmp[=argument]]"
.br
.B "[\-\-facility=syslog facility]"
.br
.B "[\-\-priority=syslog priority]"

.SH DESCRIPTION
.B iplog
is a TCP/IP traffic logger. Currently, it is capable of logging
TCP, UDP and ICMP traffic. Adding support for other protocols
should be relatively easy.
.B iplog's
capabilities include the ability to detect TCP port scans, TCP null scans,
FIN scans, UDP and ICMP "smurf" attacks,
bogus TCP flags (used by scanners to detect the operating system in use),
TCP SYN scans, TCP "Xmas" scans, ICMP ping floods, UDP scans, and IP
fragment attacks.
.B iplog
is able to run in promiscuous mode and monitor traffic to all hosts
on a network.
.B iplog
uses libpcap to read data from the network and can be ported
to any system that supports pthreads and on which libpcap will function.


.SH NOTATION
Throughout this document, required parameters will be denoted by enclosing the parameter in angle brackets <like this>.

Optional parameters will be denoted by enclosing the parameter in square brackets [like this].

The '|' character is used to express exclusive or.  For example [true|false] means you may give "true" or "false", but not both.

.SH OPTIONS
.TP

.B "\-\-tcp=true (default)"
Log TCP traffic.

.TP
.B "\-\-tcp=false"
Do not log TCP traffic.

.TP
.B "\-\-udp=true (default)"
Log UDP traffic.

.TP
.B "\-\-udp=false"
Do not log UDP traffic.

.TP
.B "\-\-icmp=true (default)"
Log ICMP traffic.

.TP
.B "\-\-icmp=false"
Do not log ICMP traffic.

.TP
.B "\-\-facility=syslog facility"
Use the specified facility for openlog(3).

.TP
.B "\-\-priority=syslog priority"
Use the specified priority for syslog(3).

.TP
.B "\-D, \-\-log\-dest=true"
Log the destination address of IP packets.

.TP
.B "\-\-log\-dest=false (default)"
Do not log the destination address of IP packets.

.TP
.B "\-F, \-\-detect\-udp\-scan=true (default)"
Detect and log UDP scans.

.TP
.B "\-\-detect\-udp\-scan=false"
Neither detect nor log UDP scans.

.TP
.B "\-\-log\-udp\-scan"
Same as \-\-detect\-udp\-scan.

.TP
.B "\-I, \-\-icmp\-resolve=true (default)"
Perform host name resolution for ICMP traffic.

.TP
.B "\-L, \-\-stdout"
Log to stdout.

.TP
.B "\-\-icmp\-resolve=false"
Do not perform host name resolution for ICMP traffic.

.TP
.B "\-N, \-\-disable\-resolver"
Do not perform host name resolution for any traffic.

.TP
.B "\-P, \-\-detect\-ping\-flood=true (default)"
Detect ping (ICMP echo) flood attacks.

.TP
.B "\-\-detect\-ping\-flood=false"
Do not detect ping flood attacks.

.TP
.B "\-\-log\-ping\-flood"
Same as \-\-detect\-ping\-flood.

.TP
.B "\-R, \-\-restart"
Restart
.B iplog,
if it is running.

.TP
.B "\-S, \-\-detect\-smurf=true (default)"
Detect "smurf" attacks.

.TP
.B "\-\-detect\-smurf=false"
Do not detect "smurf" attacks.

.TP
.B "\-\-log\-smurf"
Same as \-\-detect\-smurf.

.TP
.B "\-T, \-\-tcp\-resolve=true (default)"
Perform host name resolution for TCP traffic.

.TP
.B "\-\-tcp\-resolve=false"
Do not perform host name resolution for TCP traffic.

.TP
.B "\-U, \-\-udp\-resolve=true (default)"
Perform host name resolution for UDP traffic.

.TP
.B "\-\-udp\-resolve=false"
Do not perform host name resolution for UDP traffic.

.TP
.B "\-V, \-\-verbose=true"
Verbose - Log packets with a bad checksum and packets with a short header length.

.TP
.B "\-\-verbose=false (default)"
Do not be verbose.

.TP
.B "\-a <network,network2,...>, \-\-promisc=<network,network2,...>"
Put all monitored interfaces into promiscuous mode and log traffic destined to all hosts on the specified network(s).

.TP
.B "\-b, \-\-detect\-bogus=true (default)"
Detect bogus TCP flags.  Programs such as nmap and queso may set these flags while trying to perform OS detection.

.TP
.B "\-\-detect\-bogus=false"
Do not detect bogus TCP flags.

.TP
.B "\-\-log\-bogus"
Same as \-\-detect\-bogus.

.TP
.B "\-c, \-\-dns\-cache=true (default)"
Use a built-in DNS cache (allows host lookups to be faster).

.TP
.B "\-\-dns\-cache=false"
Do not use the built-in DNS cache.

.TP
.B "\-d, \-\-ignore"
Ignore DNS traffic from hosts listed in
.B /etc/resolv.conf.

.TP
.B "\-e, \-\-get\-ident=true"
Perform ident (RFC 1413) lookups on connections destined to a listening port.  This is only available on Linux.

.TP
.B "\-\-get\-ident=false (default)"
Do not perform ident lookups.

.TP
.B "\-f, \-\-detect\-fin\-scan=true (default)"
Detect TCP FIN scans (a "stealth scan" used by nmap and other scanners).

.TP
.B "\-\-detect\-fin\-scan=false"
Do not detect TCP FIN scans.

.TP
.B "\-\-log\-fin\-scan"
Same as \-\-detect\-fin\-scan.

.TP
.B "\-q, \-\-detect\-syn\-scan=true (default)"
Detect TCP SYN scans (a "stealth scan" used by nmap and other scanners).

.TP
.B "\-\-detect\-syn\-scan=false"
Do not detect TCP SYN scans.

.TP
.B "\-\-log\-syn\-scan"
Same as \-\-detect\-syn\-scan.


.TP
.B "\-g <group|GID>, \-\-group=<group|GID>"
Run with the specified group or GID.

.TP
.B "\-h, \-\-help"
Print a summary of available options and exit.

.TP
.B "\-i <interface(s)>, \-\-interface=<interface(s)>"
Listen on only the specified interfaces.  This option takes a comma-delimited list of interfaces.  By default,
.B iplog
will listen on any interfaces that are up, except loopback.

.TP
.B "\-k, \-\-kill"
Kill
.B iplog,
if it is running.

.TP
.B "\-l <logfile>, \-\-logfile=<logfile>"
Log to the specified file instead of logging via
.BR syslog (3)

.TP
.B "\-\-pid\-file=<file>"
Use <file> as the pid file.
.br

This option should be used when starting
.B iplog
as a user who doesn't have write access to /var/run.
.br

This option must be used with the -k and -R options when an instance of
.B iplog
is running that was started with the \-\-pid\-file option.  Also note the
\-\-pid\-file option must be given before the \-k and \-R options.

.TP
.B "\-m, \-\-scans\-only=true"
Only log scans and floods.  Do not log other traffic.

.TP
.B "\-n, \-\-detect\-null\-scan=true (default)"
Detect null scans (a "stealth scan" used by nmap and other scanners).

.TP
.B "\-\-detect\-null\-scan=false"
Do not detect null scans.

.TP
.B "\-\-log\-null\-scan"
Same as \-\-detect\-null\-scan.

.TP
.B "\-o, \-\-no\-fork"
Run in the foreground.

.TP
.B "\-p, \-\-detect\-portscan=true (default)"
Detect port scans (connect(2) scans and SYN (half open) scans).

.TP
.B "\-\-detect\-portscan=false"
Do not detect port scans.

.TP
.B "\-\-log\-portscan"
Same as \-\-detect\-portscan.

.TP
.B "\-s, \-\-detect\-syn\-flood=true (default)"
Stop resolving IP addresses (until the flood ends) if a SYN flood is detected.

.TP
.B "\-\-detect\-syn\-flood=false"
Do not stop resolving IP addresses if a SYN flood is detected.

.TP
.B "\-t, \-\-detect\-traceroute=true (default)"
Detect (and log) traceroute.

.TP
.B "\-\-detect\-traceroute=false"
Do not detect traceroute.

.TP
.B "\-\-log\-traceroute"
Same as \-\-detect\-traceroute.

.TP
.B "\-u <user|UID>, \-\-user=<user|UID>"
Run as the user or with the UID specified.

.TP
.B "\-v, \-\-version"
Print version information and exit.

.TP
.B "\-w, \-\-log-ip"
Log the IP addresses as well as the hostnames of hosts that are looked up.

.TP
.B "\-x, \-\-detect\-xmas\-scan=true (default)"
Detect Xmas scans (a "stealth" scan used by nmap and other scanners).

.TP
.B "\-\-detect-xmas-scan=false"
Do not detect Xmas scans.

.TP
.B "\-\-log\-xmas\-scan"
Same as \-\-detect\-xmas\-scan.

.TP
.B "\-y, \-\-detect\-frag=true"
Detect fragment attacks.

.TP
.B "\-\-detect\-frag=false"
Do not detect fragment attacks.

.TP
.B "\-\-log\-frag"
Same as \-\-detect\-frag.

.TP
.B "\-z, \-\-fool\-nmap=true"
Attempt to fool programs, such as nmap and queso, that perform remote OS detection.  As a side effect, this option will also cause most of nmap's "stealth" scans to fail.
.TP
.B WARNING
This option is dangerous and can set off network traffic storms.

.TP
.B "\-\-fool\-nmap=false (default)"
Do not attempt to fool nmap's OS detection.

.SH FILES
.TP
.B /etc/iplog.conf
The
.B iplog
configuration file.

.SH BUGS
Report any bugs to odin@numb.org

.SH AUTHOR
Ryan McCabe <odin@numb.org>

.SH AVAILABILITY
The primary distribution site for
.B iplog
is http://ojnk.sourceforge.net

.SH MIRROR LIST
.br
ftp://ojnk.sourceforge.net/pub/ojnk/iplog
.br
http://www.numb.org/~odin


.SH SEE ALSO
.BR iplog.conf (5)
.BR tcpdump (1)
.BR syslog (3)
.BR openlog (3)
.BR pcap (3)
.BR nmap (8)


src/iplog.c
/*
** iplog.c - iplog main routine.
** Copyright (C) 1999-2001 Ryan McCabe <odin@numb.org>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License, version 2,
** as published by the Free Software Foundation.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA
**
** $Id: iplog.c,v 1.1.1.1 2010/09/02 01:57:41 reynolds Exp $
*/

#define _GNU_SOURCE

#include <config.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <syslog.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <pthread.h>
#include <pcap.h>

#ifdef HAVE_PTHREAD_CANCEL
#	include <setjmp.h>
#endif

#ifndef __linux__
#	define THREADED_LIBPCAP_IS_BROKEN	1
#endif

#ifdef THREADED_LIBPCAP_IS_BROKEN
#	include <pcap-int.h>
#endif

#include <iplog.h>
#include <iplog_options.h>
#include <iplog_config.h>
#include <iplog_pcap.h>
#include <iplog_input.h>
#include <iplog_dns.h>
#include <iplog_scan.h>

#ifdef HAVE_PTHREAD_CANCEL
#	define setup_thread_cancelstate() \
		do { \
			pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL); \
			pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL); \
		} while (0)

	static void iplog_restart(int sig);
	static void pcap_cleanup(void *data);

	static sigjmp_buf jmpbuf;
#else
#	define	setup_thread_cancelstate() do { } while (0)
#endif

struct running {
	struct running *next;
	struct running *prev;
	pthread_t pt;
};

static struct running *running = NULL;
static struct pcap_data *plist = NULL;
static volatile bool cap_rawsock = false;
static u_int reading = 0;
static pthread_mutex_t running_lock = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t plist_lock = PTHREAD_MUTEX_INITIALIZER;

static void *read_packets(void *data);
static void *expire_data(void *unused);
static void iplog_cleanup(int sig);
static void pq_add(struct ip *ip, size_t len);
static struct packet_list *pq_get(void);
static void *queue_manager(void *data);

u_char *ifstring, *logfile, *user, *group, *lockfile;

struct packet_queue {
	struct packet_list {
		struct packet_list *next;
		struct packet_list *prev;
		struct ip *ip;
		time_t received;
	} *head, *tail;
	pthread_mutex_t lock;
	pthread_cond_t cond;
};

static struct packet_queue packet_queue = {
	head:	NULL,
	tail:	NULL,
	lock:	PTHREAD_MUTEX_INITIALIZER,
	cond:	PTHREAD_COND_INITIALIZER
};

/*
** Add a node to the packet queue.
*/

static void pq_add(struct ip *ip, size_t len) {
	struct packet_list *new_node = xmalloc(sizeof(struct packet_list));

	new_node->received = time(NULL);

	/*
	** libpcap stores packet data in a static buffer.
	** Guess what happens without the copy if there's
	** more than one interface open.
	*/

	new_node->ip = xmalloc(len);
	memcpy(new_node->ip, ip, len);

	pthread_mutex_lock(&packet_queue.lock);

	dlist_prepend(new_node, &packet_queue.head);
	if (packet_queue.tail == NULL)
		packet_queue.tail = packet_queue.head;

	pthread_cond_signal(&packet_queue.cond);
	pthread_mutex_unlock(&packet_queue.lock);
}

/*
** Remove a node from the packet queue.
** This must be called with packet_queue.lock held.
**
** The caller is responsible for freeing any space associated with the
** removed node!
*/

static struct packet_list *pq_get(void) {
	struct packet_list *old_tail = packet_queue.tail;

	if (old_tail == NULL)
		return (NULL);

	packet_queue.tail = packet_queue.tail->prev;
	dlist_remove(old_tail, &packet_queue.head);

	pthread_mutex_unlock(&packet_queue.lock);

	return (old_tail);
}

int main(int argc, char **argv) {
	struct pcap_data *cur;
	struct running rtemp;
	volatile bool dropped = false;
	int test_socket;

	lockfile = xstrdup(LOCKFILE);
	parse_config(CONFFILE);

	/* Command-line options override the defaults set in the conf file. */
	get_options(argc, argv);
	check_options();

	myopenlog("iplog", LOG_PID | LOG_NDELAY);

	if (!opt_enabled(NO_FORK))
		fork_to_back();

	write_lockfile(lockfile);

	if (opt_enabled(DNS_CACHE))
		init_dns_table(opt_enabled(PROMISC) ? DNS_MAXSIZE_P : DNS_MAXSIZE_N);

	if (opt_enabled(ANY_SCAN))
		init_scan_table(opt_enabled(PROMISC) ? SCAN_TSIZE_P : SCAN_TSIZE_N);

	init_frag_table(FRAG_TSIZE);

#ifdef HAVE_PTHREAD_CANCEL
	sigsetjmp(jmpbuf, 1);
#endif

	if (plist == NULL && setup_pcap(&plist, ifstring) == -1)
		fatal("Couldn't initialize interfaces.");

	if (dropped == false) {
		dropped = true;
		drop_privs(user, group);
	}

	memset(&rtemp, 0, sizeof(rtemp));
	pthread_create(&rtemp.pt, NULL, queue_manager, NULL);

	/* This is not the best check, but it's better than checking {e,}uid */
	test_socket = socket(PF_INET, SOCK_RAW, IPPROTO_TCP);
	if (test_socket != -1) {
		cap_rawsock = true;
		close(test_socket);
	}

	if (opt_enabled(IGNORE_NS) && add_dns_ignore_rules() == -1)
		mysyslog("Unable to add dns ignore rules: %s", strerror(errno));

	pthread_mutex_lock(&running_lock);
	pthread_mutex_lock(&plist_lock);

	pthread_create(&rtemp.pt, NULL, expire_data, NULL);
	dlist_copy_prepend(&rtemp, &running, sizeof(rtemp));

	for (cur = plist ; cur != NULL ; cur = cur->next) {
		pthread_create(&rtemp.pt, NULL, read_packets, cur);
		dlist_copy_prepend(&rtemp, &running, sizeof(rtemp));
		reading++;
	}

	pthread_mutex_unlock(&plist_lock);
	pthread_mutex_unlock(&running_lock);

	if (opt_enabled(NO_FORK))
		signal(SIGINT, iplog_cleanup);
	signal(SIGTERM, iplog_cleanup);
	signal(SIGSEGV, iplog_cleanup);
#ifdef HAVE_PTHREAD_CANCEL
	signal(SIGHUP, iplog_restart);
#else
	signal(SIGHUP, SIG_IGN);
#endif

	select(0, NULL, NULL, NULL, NULL);
	exit(0);
}

/*
** Packet queue manager thread.
** This stays alive across restarts.
*/

static void *queue_manager(void *unused) {
	struct packet_list *packet;
	sigset_t expire_sigset;

	sigemptyset(&expire_sigset);

	if (opt_enabled(NO_FORK))
		sigaddset(&expire_sigset, SIGINT);
	sigaddset(&expire_sigset, SIGTERM);
	sigaddset(&expire_sigset, SIGHUP);
	pthread_sigmask(SIG_BLOCK, &expire_sigset, NULL);

	pthread_mutex_lock(&packet_queue.lock);
	for (;;) {
		packet = pq_get();
		if (packet == NULL)
			pthread_cond_wait(&packet_queue.cond, &packet_queue.lock);
		else {
			parse_packet(packet->ip);
			free(packet->ip);
			free(packet);
			pthread_mutex_lock(&packet_queue.lock);
		}
	}

	return (unused);
}

/*
** Read in packets, send them off the appropriate function.
*/

static void *read_packets(void *data) {
	u_char *packet;
	struct pcap_pkthdr pkthdr;
	struct pcap_data *pdev = (struct pcap_data *) data;
	struct running *cur;
	sigset_t iplog_sigset;

#ifdef THREADED_LIBPCAP_IS_BROKEN
	/*
	** I'm not sure if this crap is actually needed on all
	** platforms that aren't Linux.  BSD seemed to need it, but
	** I don't know about any others (I don't have access to others
	** so I can't test) ..
	*/

	struct timeval tv;
	int fd = ((struct pcap *)pdev->pd)->fd;
	fd_set rfds;

	FD_ZERO(&rfds);
	tv.tv_sec = 0;
	tv.tv_usec = 10000;
#endif

	setup_thread_cancelstate();

	sigemptyset(&iplog_sigset);
	if (opt_enabled(NO_FORK))
		sigaddset(&iplog_sigset, SIGINT);
	sigaddset(&iplog_sigset, SIGTERM);
	sigaddset(&iplog_sigset, SIGHUP);
	pthread_sigmask(SIG_BLOCK, &iplog_sigset, NULL);

	for (;;) {
#ifdef THREADED_LIBPCAP_IS_BROKEN
		FD_SET(fd, &rfds);
		select(fd + 1, &rfds, NULL, NULL, &tv);
#endif
		packet = (u_char *) pcap_next(pdev->pd, &pkthdr);
		if (packet != NULL) {
			packet += pdev->dl;

			/* pq_add() handles locking. */
			pq_add((struct ip *) packet, pkthdr.caplen);

			if (pkthdr.caplen != pkthdr.len)
				IDEBUG(("caplen=%lu != len=%lu\n", pkthdr.caplen, pkthdr.len));
		} else if (errno == ENETDOWN) {
			mysyslog("Warning: interface %s went down.", pdev->name);
			pcap_close(pdev->pd);

			if (cap_rawsock == false) {
				mysyslog("Interface %s cannot be brought back up.", pdev->name);

				if (--reading == 0)
					fatal("No more interfaces open for reading.  Exiting.");

				pthread_mutex_lock(&running_lock);
				cur = running;

				/* Can't ever hit NULL */
				while (cur->pt != pthread_self())
					cur = cur->next;

				dlist_delete(cur, &running);
				pthread_mutex_unlock(&running_lock);

				pthread_mutex_lock(&plist_lock);
				dlist_delete(pdev, &plist);
				pthread_mutex_unlock(&plist_lock);

				pthread_exit(NULL);
			} else {
				struct pcap_data *dplist = NULL;

				for (;;) {
					if (setup_pcap(&dplist, pdev->name) != -1) {
						if (ifflag_isset(dplist->name, IFF_UP)) {
							mysyslog("Interface %s was reopened.",
								dplist->name);

							pthread_mutex_lock(&plist_lock);
							dlist_delete(pdev, &plist);
							pdev = dlist_prepend(dplist, &plist);
							pthread_mutex_unlock(&plist_lock);
							break;
						} else {
							pcap_close(dplist->pd);
							dlist_destroy(dplist, NULL);
						}
					}

					/* Try again in 30 seconds. */
					xsleep(30);
				}
			}
		}
	}

	return (NULL);
}

/*
** This function runs as a separate thread, expiring fragment data,
** scan hash table entries and dns hash table entries.
*/

static void *expire_data(void *unused) {
	sigset_t expire_sigset;

	setup_thread_cancelstate();
	sigemptyset(&expire_sigset);

	if (opt_enabled(NO_FORK))
		sigaddset(&expire_sigset, SIGINT);
	sigaddset(&expire_sigset, SIGTERM);
	sigaddset(&expire_sigset, SIGHUP);
	pthread_sigmask(SIG_BLOCK, &expire_sigset, NULL);

	for (;;) {
		xsleep(EXPIRE_INTERVAL);

		expire_frags();
		if (opt_enabled(ANY_SCAN))
			expire_scans();

		if (opt_enabled(DNS_CACHE))
			expire_dns();
	}

	return (unused);
}

#ifdef HAVE_PTHREAD_CANCEL

/*
** Pcap cleanup function.
*/

static void pcap_cleanup(void *data) {
	struct pcap_data *cur = data;

	pcap_close(cur->pd);
}

/*
** Restart iplog.  Intended to be a signal handler.
*/

static void iplog_restart(int sig) {
	struct running *cur;
	extern pthread_mutex_t log_lock;

	pthread_mutex_lock(&packet_queue.lock);
	xusleep(2000);
	pthread_mutex_lock(&log_lock);

	for (cur = running ; cur != NULL ;) {
		pthread_cancel(cur->pt);
		pthread_join(cur->pt, NULL);
		cur = dlist_delete(cur, &running);
	}

	reading = 0;

	if (opt_enabled(ANY_SCAN))
		destroy_scan_table();

	if (opt_enabled(DNS_CACHE))
		destroy_dns_cache();

	destroy_frag_table();

	if (cap_rawsock == true) {
		dlist_destroy(plist, pcap_cleanup);
		plist = NULL;
	}

	destroy_filter_list(FIL_TCP);
	destroy_filter_list(FIL_UDP);
	destroy_filter_list(FIL_ICMP);

	pthread_mutex_unlock(&log_lock);
	pthread_mutex_unlock(&packet_queue.lock);

	mysyslog("Restarting iplog (pid %d).", getpid());
	mycloselog();

    parse_config(CONFFILE);
	check_options();
	myopenlog("iplog", LOG_PID | LOG_NDELAY);

	siglongjmp(jmpbuf, 1);

	/* Quiet gcc */
	(void) sig;
}
#endif

/*
** Iplog cleanup function.
*/

static void iplog_cleanup(int sig) {
	if (unlink(lockfile) == -1)
		mysyslog("Couldn't unlink \"%s\": %s", lockfile, strerror(errno));
	mysyslog("Caught signal %d, exiting.", sig);
	exit(0);
}

/* vim:ts=4:sw=8:tw=0 */
